File: data_structure/tree/src/TreeDemo.go

// Go
// データ構造: 木 (Tree)

package main

import (
	"fmt"
)

type NodeData struct {
	value    interface{}
	parent   *NodeData
	children []*NodeData
}

func NewNodeData(value interface{}) *NodeData {
	return &NodeData{
		value:    value,
		parent:   nil,
		children: []*NodeData{},
	}
}

func (n *NodeData) GetValue() interface{} {
	return n.value
}

func (n *NodeData) GetParent() *NodeData {
	return n.parent
}

func (n *NodeData) GetChildren() []*NodeData {
	return n.children
}

func (n *NodeData) SetParent(parent *NodeData) bool {
	n.parent = parent
	return true
}

func (n *NodeData) AddChild(child *NodeData) bool {
	child.SetParent(n)
	n.children = append(n.children, child)
	return true
}

func (n *NodeData) RemoveChild(child *NodeData) bool {
	for i, c := range n.children {
		if c == child {
			child.SetParent(nil)
			n.children = append(n.children[:i], n.children[i+1:]...)
			return true
		}
	}
	return false
}

func (n *NodeData) IsLeaf() bool {
	return len(n.children) == 0
}

type TreeData struct {
	data *NodeData
}

func NewTreeData() *TreeData {
	return &TreeData{
		data: nil,
	}
}

func (t *TreeData) Get() *NodeData {
	return t.data
}

func (t *TreeData) GetHeight(node *NodeData) int {
	if node == nil {
		node = t.data
	}
	if node == nil {
		return 0
	}
	if node.IsLeaf() {
		return 1
	}
	maxHeight := 0
	for _, child := range node.GetChildren() {
		height := t.GetHeight(child)
		if height > maxHeight {
			maxHeight = height
		}
	}
	return 1 + maxHeight
}

func (t *TreeData) GetParent(node *NodeData) *NodeData {
	return node.GetParent()
}

func (t *TreeData) GetChildren(node *NodeData) []*NodeData {
	return node.GetChildren()
}

func (t *TreeData) GetNode(value interface{}, node *NodeData) *NodeData {
	if node == nil {
		node = t.data
	}
	if node == nil {
		return nil
	}

	if node.GetValue() == value {
		return node
	}

	for _, child := range node.GetChildren() {
		result := t.GetNode(value, child)
		if result != nil {
			return result
		}
	}

	return nil
}

func (t *TreeData) Add(parent *NodeData, value interface{}) bool {
	newNode := NewNodeData(value)
	if parent == nil {
		if t.data == nil {
			t.data = newNode
			return true
		} else {
			fmt.Printf("ERROR: %v 重複です\n", value)
			return false
		}
	} else {
		parent.AddChild(newNode)
		return true
	}
}

func (t *TreeData) Remove(node *NodeData) bool {
	if node == nil {
		return false
	}

	if node == t.data {
		t.data = nil
		return true
	}

	parent := node.GetParent()
	if parent != nil {
		return parent.RemoveChild(node)
	}

	return false
}

func (t *TreeData) Traverse(node *NodeData, mode string) []interface{} {
	if node == nil {
		node = t.data
	}
	if node == nil {
		return []interface{}{}
	}

	result := []interface{}{}

	switch mode {
	case "pre-order":
		result = append(result, node.GetValue())
		for _, child := range node.GetChildren() {
			result = append(result, t.Traverse(child, mode)...)
		}
	case "post-order":
		for _, child := range node.GetChildren() {
			result = append(result, t.Traverse(child, mode)...)
		}
		result = append(result, node.GetValue())
	case "level-order":
		queue := []*NodeData{node}
		for len(queue) > 0 {
			current := queue[0]
			queue = queue[1:]
			result = append(result, current.GetValue())
			queue = append(queue, current.GetChildren()...)
		}
	}

	return result
}

func (t *TreeData) IsLeaf(node *NodeData) bool {
	return node != nil && node.IsLeaf()
}

func (t *TreeData) IsEmpty() bool {
	return t.data == nil
}

func (t *TreeData) Size(node *NodeData) int {
	if node == nil {
		node = t.data
	}
	if node == nil {
		return 0
	}

	count := 1
	for _, child := range node.GetChildren() {
		count += t.Size(child)
	}

	return count
}

func (t *TreeData) Clear() bool {
	t.data = nil
	return true
}

func (t *TreeData) Display() []interface{} {
	if t.data == nil {
		return []interface{}{}
	}

	return t.Traverse(nil, "level-order")
}

func main() {
	fmt.Println("Tree TEST -----> start")

	fmt.Println("\nnew")
	treeData := NewTreeData()
	fmt.Printf("  現在のデータ: %v\n", treeData.Display())

	fmt.Println("\nis_empty")
	output := treeData.IsEmpty()
	fmt.Printf("  出力値: %v\n", output)

	fmt.Println("\nsize")
	output2 := treeData.Size(nil)
	fmt.Printf("  出力値: %v\n", output2)

	fmt.Println("\nadd")
	fmt.Printf("  入力値: %v, %v\n", nil, "Root")
	output3 := treeData.Add(nil, "Root")
	fmt.Printf("  出力値: %v\n", output3)
	fmt.Printf("  現在のデータ: %v\n", treeData.Display())

	rootNode := treeData.Get()

	fmt.Println("\nadd")
	fmt.Printf("  入力値: %v, %v\n", rootNode, "Child1")
	output4 := treeData.Add(rootNode, "Child1")
	fmt.Printf("  出力値: %v\n", output4)
	fmt.Printf("  現在のデータ: %v\n", treeData.Display())

	fmt.Println("\nadd")
	fmt.Printf("  入力値: %v, %v\n", rootNode, "Child2")
	output5 := treeData.Add(rootNode, "Child2")
	fmt.Printf("  出力値: %v\n", output5)
	fmt.Printf("  現在のデータ: %v\n", treeData.Display())

	fmt.Println("\nget_node")
	inputValue := "Child1"
	fmt.Printf("  入力値: %v\n", inputValue)
	output6 := treeData.GetNode(inputValue, nil)
	fmt.Printf("  出力値: %v\n", output6)
	fmt.Printf("  現在のデータ: %v\n", treeData.Display())

	fmt.Println("\ntraverse")
	inputMode := "pre-order"
	fmt.Printf("  入力値: %v\n", inputMode)
	output7 := treeData.Traverse(nil, inputMode)
	fmt.Printf("  出力値: %v\n", output7)
	fmt.Printf("  現在のデータ: %v\n", treeData.Display())

	fmt.Println("\nTree TEST <----- end")
}

File: data_structure/tree/src/TreeDemo.java
// Java
// データ構造: 木 (Tree)

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class NodeData {
    private Object value;
    private NodeData parent;
    private List<NodeData> children;
    
    public NodeData(Object value) {
        this.value = value;
        this.parent = null;
        this.children = new ArrayList<>();
    }
    
    public Object getValue() {
        return value;
    }
    
    public NodeData getParent() {
        return parent;
    }
    
    public List<NodeData> getChildren() {
        return children;
    }
    
    public boolean setParent(NodeData parent) {
        this.parent = parent;
        return true;
    }
    
    public boolean addChild(NodeData child) {
        child.setParent(this);
        children.add(child);
        return true;
    }
    
    public boolean removeChild(NodeData child) {
        if (children.contains(child)) {
            child.setParent(null);
            children.remove(child);
            return true;
        } else {
            return false;
        }
    }
    
    public boolean isLeaf() {
        return children.isEmpty();
    }
}

class TreeData {
    private NodeData data;
    
    public TreeData() {
        this.data = null;
    }
    
    public NodeData get() {
        return data;
    }
    
    public int getHeight(NodeData node) {
        if (node == null) {
            node = data;
        }
        if (node == null) {
            return 0;
        }
        if (node.isLeaf()) {
            return 1;
        }
        
        int maxHeight = 0;
        for (NodeData child : node.getChildren()) {
            int height = getHeight(child);
            if (height > maxHeight) {
                maxHeight = height;
            }
        }
        return 1 + maxHeight;
    }
    
    public NodeData getParent(NodeData node) {
        return node.getParent();
    }
    
    public List<NodeData> getChildren(NodeData node) {
        return node.getChildren();
    }
    
    public NodeData getNode(Object value, NodeData node) {
        if (node == null) {
            node = data;
        }
        if (node == null) {
            return null;
        }
        
        if (node.getValue().equals(value)) {
            return node;
        }
        
        for (NodeData child : node.getChildren()) {
            NodeData result = getNode(value, child);
            if (result != null) {
                return result;
            }
        }
        
        return null;
    }
    
    public NodeData getNode(Object value) {
        return getNode(value, null);
    }
    
    public boolean add(NodeData parent, Object value) {
        NodeData newNode = new NodeData(value);
        if (parent == null) {
            if (data == null) {
                data = newNode;
                return true;
            } else {
                System.out.println("ERROR: " + value + " 重複です");
                return false;
            }
        } else {
            return parent.addChild(newNode);
        }
    }
    
    public boolean remove(NodeData node) {
        if (node == null) {
            return false;
        }
        
        if (node == data) {
            data = null;
            return true;
        }
        
        NodeData parent = node.getParent();
        if (parent != null) {
            return parent.removeChild(node);
        }
        
        return false;
    }
    
    public List<Object> traverse(NodeData node, String mode) {
        if (node == null) {
            node = data;
        }
        if (node == null) {
            return new ArrayList<>();
        }
        
        List<Object> result = new ArrayList<>();
        
        if ("pre-order".equals(mode)) {
            result.add(node.getValue());
            for (NodeData child : node.getChildren()) {
                result.addAll(traverse(child, mode));
            }
        } else if ("post-order".equals(mode)) {
            for (NodeData child : node.getChildren()) {
                result.addAll(traverse(child, mode));
            }
            result.add(node.getValue());
        } else if ("level-order".equals(mode)) {
            Queue<NodeData> queue = new LinkedList<>();
            queue.add(node);
            while (!queue.isEmpty()) {
                NodeData current = queue.poll();
                result.add(current.getValue());
                queue.addAll(current.getChildren());
            }
        }
        
        return result;
    }
    
    public List<Object> traverse(String mode) {
        return traverse(null, mode);
    }
    
    public boolean isLeaf(NodeData node) {
        return node != null && node.isLeaf();
    }
    
    public boolean isEmpty() {
        return data == null;
    }
    
    public int size(NodeData node) {
        if (node == null) {
            node = data;
        }
        if (node == null) {
            return 0;
        }
        
        int count = 1;
        for (NodeData child : node.getChildren()) {
            count += size(child);
        }
        
        return count;
    }
    
    public int size() {
        return size(null);
    }
    
    public boolean clear() {
        data = null;
        return true;
    }
    
    public List<Object> display() {
        if (data == null) {
            return new ArrayList<>();
        }
        
        return traverse("level-order");
    }
}

public class TreeDemo {
    public static void main(String[] args) {
        System.out.println("Tree TEST -----> start");
        
        System.out.println("\nnew");
        TreeData treeData = new TreeData();
        System.out.println("  現在のデータ: " + treeData.display());
        
        System.out.println("\nis_empty");
        boolean output1 = treeData.isEmpty();
        System.out.println("  出力値: " + output1);
        
        System.out.println("\nsize");
        int output2 = treeData.size();
        System.out.println("  出力値: " + output2);
        
        System.out.println("\nadd");
        Object[] inputParams1 = {null, "Root"};
        System.out.println("  入力値: " + java.util.Arrays.toString(inputParams1));
        boolean output3 = treeData.add(null, "Root");
        System.out.println("  出力値: " + output3);
        System.out.println("  現在のデータ: " + treeData.display());
        
        NodeData rootNode = treeData.get();
        
        System.out.println("\nadd");
        Object[] inputParams2 = {rootNode, "Child1"};
        System.out.println("  入力値: " + java.util.Arrays.toString(inputParams2));
        boolean output4 = treeData.add(rootNode, "Child1");
        System.out.println("  出力値: " + output4);
        System.out.println("  現在のデータ: " + treeData.display());
        
        System.out.println("\nadd");
        Object[] inputParams3 = {rootNode, "Child2"};
        System.out.println("  入力値: " + java.util.Arrays.toString(inputParams3));
        boolean output5 = treeData.add(rootNode, "Child2");
        System.out.println("  出力値: " + output5);
        System.out.println("  現在のデータ: " + treeData.display());
        
        System.out.println("\nget_node");
        String inputValue = "Child1";
        System.out.println("  入力値: " + inputValue);
        NodeData output6 = treeData.getNode(inputValue);
        System.out.println("  出力値: " + output6);
        System.out.println("  現在のデータ: " + treeData.display());
        
        System.out.println("\ntraverse");
        String inputMode = "pre-order";
        System.out.println("  入力値: " + inputMode);
        List<Object> output7 = treeData.traverse(inputMode);
        System.out.println("  出力値: " + output7);
        System.out.println("  現在のデータ: " + treeData.display());
        
        System.out.println("\nTree TEST <----- end");
    }
}